pilot/ARCHITECTURE.md

```diff
@@ -0,0 +1,163 @@
+This document describes the high level architecture of the Istio control plane, Istiod.
+Istiod is structured as a modular monolith, housing a wide range of functionality from certificate signing, proxy configuration (XDS), traditional Kubernetes controllers, and more.
+
+## Proxy Configuration
+Istiod's primary role - and most code - is to dynamically configure proxies (Envoy sidecars and ingress, gRPC, ztunnel, and more). This roughly consists of 3 parts:
+1. Config ingestion (inputs to the system)
+2. Config translation
+3. Config serving (XDS)
+### Config Ingestion
+
+Istio reads from over 20 different resources types, and aggregates them together to build the proxy configuration. These resources can be sourced from Kubernetes (via watches), files, or over xDS; Kubernetes is by far the most common usage, though.
+
+Primarily for historical reasons, ingestion is split into a few components.
+
+#### ConfigStore
+
+The `ConfigStore` reads a variety of resources and exposes them over a standard interface (Get, List, etc). These types are wrapped in a common `config.Config` struct, contrasting with typical Kubernetes clients which use per-resource types. The most common is reading from Kubernetes via the `crdclient` package.
+
+```mermaid
+graph TD
+	subgraph ConfigStore
+		xcs(XDS Client)
+		ccs(CRD Client)
+		fcs(Filesystem Client)
+		acs(Aggregate)
+		xcs-->acs
+		ccs-->acs
+		fcs-->acs
+	end
+```
+#### ServiceDiscovery
+
+The other primary interface is the ServiceDiscovery. Similar to ConfigStore, this aggregates over a variety of resources. However, it does not provide generic resource access, and instead precomputes a variety of service-oriented internal resources, such as `model.Service` and `model.ServiceInstance`.
+
+This is composed of two controllers - one driven from core Kubernetes types ("Kube Controller") and one by Istio types ("ServiceEntry controller").
+
+```mermaid
+graph TD
+	subgraph Kube Controller
+		s(Services)
+		e(Endpoints)
+		p(Pods)
+		ksi(ServiceInstances)
+		kwi(WorkloadInstances)
+		s-->ksi
+		e-->ksi
+		p-->kwi
+		
+	end
+	subgraph ServiceEntry Controller
+		se(ServiceEntry)
+		we(WorkloadEntry)
+		ssi(ServiceInstances)
+		swi(WorkloadInstances)
+		se-->ssi
+		swi-->ssi
+		we-->swi
+	end
+	kwi-->ssi
+	swi-->ksi
+```
+For the most part this is fairly straight forward. However, we support `ServiceEntry` selecting `Pod`, and `Service` selecting `WorkloadEntry`, which leads to cross-controller communication.
```

Comment:
```suggestion
For the most part this is fairly straight forward. However, we support `ServiceEntry` selecting `Pod`, and `Service` selecting `WorkloadEntry`, which leads to cross-controller communication.
```

(mainly going by the diagram here) Some more detail about _why_ these parallel controllers affect each other's _State Of The World_ (if I'm reading this right and they do), and why we don't have something more compositional, where the `KubeController` is purely tracking Kube types and remains ignorant of Istio types/side effects would be good.

i.e. "why don't ALL the arrows flow from KubeController to ServiceEntryController"

Reply:
Because it was implemented poorly is probably the best answer I can give... :slightly_smiling_face:.

Reply:
That's correct, but one of the goals was to treat things related to LB and routing differently from policies.

If I remember correctly, the first/original Istio had the config generation in agent (saving yaml files). Kuat may remember better.